docs/proposals/kubectl-extension.md

```diff
@@ -0,0 +1,85 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<!-- TAG RELEASE_LINK, added by the munger automatically -->
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.3/docs/proposals/kubectl-extension.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Kubectl Extension
+
+Abstract
+--------
+
+Allow `kubectl` to be extended to include other commands that can provide new functionality without recompiling Kubectl
+
+
+Motivation and Background
+-------------------------
+
+Kubernetes is designed to be a composable and extensible system, with the ability to add new APIs and features via Third Party Resources
+or API federation, by making the server provide functionality that eases writing generic clients, and by supporting other authentication
+systems. Given that `kubectl` is the primary method for interacting with the server, some new extensions are difficult to make usable
+for end users without recompiling that command. In addition, it is difficult to prototype new functionality for kubectl outside of the
+Kubernetes source tree.
+
+Ecosystem tools like OpenShift, Deis, and Helm add additional workflow around kubectl targeted at the end user. It is beneficial
```

Comment:
Discovery and help text is certainly the most prevalent reason.  Typically
extension is understood as best effort - the extension provider may be the
distributor of the Kube binaries, an IT organization, or a team looking to
standardize, but the addition of new interesting commands into the command
space by the project maintainers is generally given priority.

Version skew is typically a distribution problem.  Binary distribution is
effectively a solved problem in the Linux and Mac space, so I don't think
binary distribution is something this proposal should attempt solve (and I
would argue that inventing our own new and innovative way of distributing
software is not a priority for this project right now).

I only think a subset of clients need to be statically linked.  It's just
not a normal pattern on either Linux or Mac, and Windows has its own
solutions in that space.  I think extenders should be free to write in
Ruby, Python, or their language of choice (even Bash).

The smaller version of this proposal is that extension authors write
'kubectl-*' symlinks and then rely on tab completion for discovery, which
is totally reasonable.


On Mon, Aug 8, 2016 at 8:04 PM, Brian Grant <notifications@github.com>
wrote:

> In docs/proposals/kubectl-extension.md
> <https://github.com/kubernetes/kubernetes/pull/30086#discussion_r73978587>
> :
>
> > +Abstract
> > +--------
> > +
> > +Allow `kubectl` to be extended to include other commands that can provide new functionality without recompiling Kubectl
> > +
> > +
> > +Motivation and Background
> > +-------------------------
> > +
> > +Kubernetes is designed to be a composable and extensible system, with the ability to add new APIs and features via Third Party Resources
> > +or API federation, by making the server provide functionality that eases writing generic clients, and by supporting other authentication
> > +systems. Given that `kubectl` is the primary method for interacting with the server, some new extensions are difficult to make usable
> > +for end users without recompiling that command. In addition, it is difficult to prototype new functionality for kubectl outside of the
> > +Kubernetes source tree.
> > +
> > +Ecosystem tools like OpenShift, Deis, and Helm add additional workflow around kubectl targeted at the end user. It is beneficial
>
> It would be useful to describe the benefit of prefixing a command with
> "kubectl" as opposed to executing it directly, such as minikube... vs kubectl
> minikube..., kompose... vs kubectl kompose..., helm... vs. kubectl helm...,
> and so on.
>
> Controlling version skew? Discovery via kubectl help?
>
> It doesn't solve the binary distribution problem, at least not on its own.
> And I assume all the clients will either need to be statically linked or
> packaged in container images.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30086/files/b013f3367d22b4e16e3347d4e1e9ab7c2c0ea56a#r73978587>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p6kylc_9Vl6dWaSn-uL3-Ir3PFkbks5qd8P5gaJpZM4Jc6R_>
> .
>